tokens/service.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (C) 2021 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package tokens
+
+import (
+	"context"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+
+	"storj.io/storjscan/tokens/erc20"
+)
+
+// ErrService - tokens service error class.
+var ErrService = errs.Class("tokens service")
+
+// Config holds tokens service configuration.
+type Config struct {
+	Endpoint     string
+	TokenAddress string
+}
+
+// Service for querying ERC20 token information from ethereum chain.
+//
+// architecture: Service
+type Service struct {
+	log      *zap.Logger
+	endpoint string
+	token    Address
+}
+
+// NewService creates new token service instance.
+func NewService(log *zap.Logger, endpoint string, token Address) *Service {
+	return &Service{
+		log:      log,
+		endpoint: endpoint,
+		token:    token,
+	}
+}
+
+// Payments retrieves all ERC20 token payments for ethereum address.
+func (service *Service) Payments(ctx context.Context, address Address) (_ []Payment, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	client, err := ethclient.DialContext(ctx, service.endpoint)
+	if err != nil {
+		return nil, err
+	}
+	defer client.Close()
+
+	token, err := erc20.NewErc20(service.token, client)
+	if err != nil {
+		return nil, ErrService.Wrap(err)
+	}
+
+	opts := &bind.FilterOpts{
+		Start:   0,
```

Comment:
I am not sure how the results of this method will be used. But there are cases when HEAD - 3..6 can be more safe (ignoring blocks without enough confirmation).

Especially as I don't see that the `log.Removed` is handled by the generated code:

```
// Log represents a contract log event. These events are generated by the LOG opcode and
// stored/indexed by the node.
type Log struct {
	// The Removed field is true if this log was reverted due to a chain reorganisation.
	// You must pay attention to this field if you receive logs through a filter query.
	Removed bool `json:"removed"`
}
```

Do you know if it's possible to get removed log entries?

Reply:
Based on the docs when you create a subscription for new logs if the log that was sent before is removed due to blockchain reorg it will be send again with `Removed` set to `true`